From 03c1ce2001b7b6eeffa4274f70bcb3eaad2bd59c Mon Sep 17 00:00:00 2001
From: Hou Qi <qi.hou@nxp.com>
Date: Thu, 9 May 2024 15:08:21 +0800
Subject: [PATCH 65/75] MMFMWK-9338 v4l2: Make sure output is queued before
 starting src task

For streams that need orphaning after first seek, capture buffer
pool is streamoff and inactive, while output is streamon. After
second seek, output streamoff and has no chance to queue buffer
before starting src task as output buffer pool is still active.
Then capture queue try to poll when both queues are empty, which
causes poll error.

To fix this, need to make sure output buffer is queued before
starting src task, but not to wait source change event after seek
to avoid hang.

Upstream-Status: Pending
Part-of: https://gitlab.freedesktop.org/gstreamer/gstreamer/-/merge_requests/6236
Signed-off-by: Xiaolei Wang <xiaolei.wang@windriver.com>
---
 sys/v4l2/gstv4l2bufferpool.c |  8 ++++++
 sys/v4l2/gstv4l2object.c     |  1 +
 sys/v4l2/gstv4l2object.h     |  1 +
 sys/v4l2/gstv4l2videodec.c   | 49 ++++++++++++++++++++----------------
 4 files changed, 37 insertions(+), 22 deletions(-)

diff --git a/sys/v4l2/gstv4l2bufferpool.c b/sys/v4l2/gstv4l2bufferpool.c
index 91ef3e9..a78329c 100644
--- a/sys/v4l2/gstv4l2bufferpool.c
+++ b/sys/v4l2/gstv4l2bufferpool.c
@@ -707,6 +707,10 @@ gst_v4l2_buffer_pool_streamon (GstV4l2BufferPool * pool)
       if (obj->ioctl (pool->video_fd, VIDIOC_STREAMON, &obj->type) < 0)
         goto streamon_failed;
 
+      /* finish seek via OUTPUT streamon */
+      if (V4L2_TYPE_IS_OUTPUT (obj->type))
+        g_atomic_int_set (&obj->seek, FALSE);
+
       pool->streaming = TRUE;
 
       GST_DEBUG_OBJECT (pool, "Started streaming");
@@ -746,6 +750,10 @@ gst_v4l2_buffer_pool_streamoff (GstV4l2BufferPool * pool)
         GST_WARNING_OBJECT (pool, "STREAMOFF failed with errno %d (%s)",
             errno, g_strerror (errno));
 
+      /* start seek via OUTPUT streamoff */
+      if (V4L2_TYPE_IS_OUTPUT (obj->type))
+        g_atomic_int_set (&obj->seek, TRUE);
+
       pool->streaming = FALSE;
 
       GST_DEBUG_OBJECT (pool, "Stopped streaming");
diff --git a/sys/v4l2/gstv4l2object.c b/sys/v4l2/gstv4l2object.c
index cc7e86c..0f216c0 100644
--- a/sys/v4l2/gstv4l2object.c
+++ b/sys/v4l2/gstv4l2object.c
@@ -583,6 +583,7 @@ gst_v4l2_object_new (GstElement * element,
   v4l2object->can_poll_device = TRUE;
   v4l2object->crop_width = 0;
   v4l2object->crop_height = 0;
+  g_atomic_int_set (&v4l2object->seek, FALSE);
 
   /* We now disable libv4l2 by default, but have an env to enable it. */
 #ifdef HAVE_LIBV4L2
diff --git a/sys/v4l2/gstv4l2object.h b/sys/v4l2/gstv4l2object.h
index 82ed45b..c17a012 100644
--- a/sys/v4l2/gstv4l2object.h
+++ b/sys/v4l2/gstv4l2object.h
@@ -239,6 +239,7 @@ struct _GstV4l2Object {
   gboolean is_hantro;
   gint crop_width;
   gint crop_height;
+  gboolean seek;
 
   guint max_width;
   guint max_height;
diff --git a/sys/v4l2/gstv4l2videodec.c b/sys/v4l2/gstv4l2videodec.c
index c037156..00e1614 100644
--- a/sys/v4l2/gstv4l2videodec.c
+++ b/sys/v4l2/gstv4l2videodec.c
@@ -1136,6 +1136,7 @@ gst_v4l2_video_dec_handle_frame (GstVideoDecoder * decoder,
   gboolean processed = FALSE;
   GstBuffer *tmp;
   GstTaskState task_state;
+  gboolean seek = g_atomic_int_get (&self->v4l2output->seek);
 
   GST_DEBUG_OBJECT (self, "Handling frame %d", frame->system_frame_number);
 
@@ -1151,7 +1152,8 @@ gst_v4l2_video_dec_handle_frame (GstVideoDecoder * decoder,
   }
 
   pool = gst_v4l2_object_get_buffer_pool (self->v4l2output);
-  if (G_UNLIKELY (!gst_buffer_pool_is_active (pool))) {
+  if (G_UNLIKELY (!gst_buffer_pool_is_active (pool)
+          || !GST_V4L2_BUFFER_POOL_IS_STREAMING (pool))) {
     GstBuffer *codec_data;
     GstStructure *config = gst_buffer_pool_get_config (pool);
     guint min = MAX (self->v4l2output->min_buffers,
@@ -1174,32 +1176,34 @@ gst_v4l2_video_dec_handle_frame (GstVideoDecoder * decoder,
     }
 
     /* Ensure input internal pool is active */
+    if (!gst_buffer_pool_is_active (pool)) {
+      gst_buffer_pool_config_set_params (config, self->input_state->caps,
+          self->v4l2output->info.size, min, max);
+
+      /* There is no reason to refuse this config */
+      if (!gst_buffer_pool_set_config (pool, config)) {
+        config = gst_buffer_pool_get_config (pool);
+
+        if (!gst_buffer_pool_config_validate_params (config,
+                self->input_state->caps, self->v4l2output->info.size, min,
+                max)) {
+          gst_structure_free (config);
+          goto activate_failed;
+        }
 
-    gst_buffer_pool_config_set_params (config, self->input_state->caps,
-        self->v4l2output->info.size, min, max);
-
-    /* There is no reason to refuse this config */
-    if (!gst_buffer_pool_set_config (pool, config)) {
-      config = gst_buffer_pool_get_config (pool);
-
-      if (!gst_buffer_pool_config_validate_params (config,
-              self->input_state->caps, self->v4l2output->info.size, min, max)) {
-        gst_structure_free (config);
-        goto activate_failed;
+        if (!gst_buffer_pool_set_config (pool, config))
+          goto activate_failed;
       }
 
-      if (!gst_buffer_pool_set_config (pool, config))
+      /* Ensure to unlock capture, as it may be flushing due to previous
+      * unlock/stop calls */
+      gst_v4l2_object_unlock_stop (self->v4l2output);
+      gst_v4l2_object_unlock_stop (self->v4l2capture);
+
+      if (!gst_buffer_pool_set_active (pool, TRUE))
         goto activate_failed;
     }
 
-    /* Ensure to unlock capture, as it may be flushing due to previous
-     * unlock/stop calls */
-    gst_v4l2_object_unlock_stop (self->v4l2output);
-    gst_v4l2_object_unlock_stop (self->v4l2capture);
-
-    if (!gst_buffer_pool_set_active (pool, TRUE))
-      goto activate_failed;
-
     GST_VIDEO_DECODER_STREAM_UNLOCK (decoder);
     GST_LOG_OBJECT (decoder, "Passing buffer with system frame number %u",
         processed ? frame->system_frame_number : 0);
@@ -1219,7 +1223,8 @@ gst_v4l2_video_dec_handle_frame (GstVideoDecoder * decoder,
     gst_buffer_unref (codec_data);
 
     /* Only wait for source change if the formats supports it */
-    if (!GST_V4L2_IS_ACTIVE (self->v4l2capture) &&
+    /* No need to wait for source change after seek */
+    if (!GST_V4L2_IS_ACTIVE (self->v4l2capture) && !seek &&
         self->v4l2output->fmtdesc->flags & V4L2_FMT_FLAG_DYN_RESOLUTION) {
       gst_v4l2_object_unlock_stop (self->v4l2capture);
       self->wait_for_source_change = TRUE;
-- 
2.43.0

